pkg/container-collection/tracerconnect/connect.go

```diff
@@ -0,0 +1,122 @@
+
+import (
+	"fmt"
+	"os"
+
+	containercollection "github.com/inspektor-gadget/inspektor-gadget/pkg/container-collection"
+	containerutils "github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils"
+	eventtypes "github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+)
+
+type NetworkingTracer[Event any] interface {
+	Attach(key string, pid uint32, eventCallback func(Event)) error
+	Detach(key string) error
+}
+
+type ConnectNetworkingTracerToContainerCollectionArgs[Event any] struct {
+	Key           string
+	Tracer        NetworkingTracer[Event]
+	Resolver      containercollection.ContainerResolver
+	Selector      containercollection.ContainerSelector
+	EventCallback func(*containercollection.Container, Event)
+	Base          func(eventtypes.Event) Event
+}
+
+// ConnectNetworkingTracerToContainerCollection connects a networking
+// tracer to the container collection package. It creates the needed
+// logic to call the Attach() function on the tracer each time a
+// container is created and to call Detach() each time the container is
+// removed. Callers must call args.Resolver.Unsubscribe(key) to remove
+// the connection.
+func ConnectNetworkingTracerToContainerCollection[Event any](
+	args *ConnectNetworkingTracerToContainerCollectionArgs[Event],
+) error {
+	return connectNetworkingTracerToContainerCollectionInternal(
+		args.Key, args.Tracer, args.Resolver, args.Selector, args.EventCallback,
+		args.Base,
+	)
+}
+
+func connectNetworkingTracerToContainerCollectionInternal[Event any](
+	key string,
+	tracer NetworkingTracer[Event],
+	resolver containercollection.ContainerResolver,
+	selector containercollection.ContainerSelector,
+	eventCallback func(*containercollection.Container, Event),
+	base func(eventtypes.Event) Event,
+) error {
+	netnsHost, err := containerutils.GetNetNs(os.Getpid())
+	if err != nil {
+		return fmt.Errorf("getting network namespace: %w", err)
+	}
+	genKey := func(container *containercollection.Container) string {
+		if container.Netns == netnsHost {
+			return "host"
+		}
+
+		key := container.Namespace + container.Podname
+		// for local-gadget as container.Namespace and
+		// container.Podname can be empty
+		if key == "" {
+			key = container.ID
+		}
+
+		return key
+	}
+
+	attachContainerFunc := func(container *containercollection.Container) {
+		cbWithContainer := func(ev Event) {
+			eventCallback(container, ev)
+		}
+		err := tracer.Attach(genKey(container), container.Pid, cbWithContainer)
+		if err != nil {
+			msg := fmt.Sprintf("failed to trace container: %q: %s", container.Name, err)
+			cbWithContainer(base(eventtypes.Err(msg)))
+			return
+		}
+		cbWithContainer(base(eventtypes.Debug("tracer attached")))
+	}
+
+	detachContainerFunc := func(container *containercollection.Container) {
+		err := tracer.Detach(genKey(container))
+		if err != nil {
+			msg := fmt.Sprintf("failed to stop tracing container: %q: %s", container.Name, err)
+			eventCallback(container, base(eventtypes.Err(msg)))
```

Comment:
Line 89 and 99 could use the same way to call the callback for consistency:
Either
```
cbWithContainer(base(eventtypes.Err(msg)))
```
or
```
eventCallback(container, base(eventtypes.Err(msg)))
```